File: solutions/Python/remove_nth_node_from_end_of_list_two_pass.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def remove_nth_node_two_pass(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """
    Remove the nth node from the end of a linked list using two passes.

    Approach:
    1. First pass: count the total number of nodes
    2. Second pass: find and skip the nth node from the end

    Time: O(L) + O(L-n) = O(L)
    Space: O(1)
    """
    # First pass: count the total nodes
    length = 0
    current = head
    while current:
        length += 1
        current = current.next

    # Edge case: removing the head node
    if length == n:
        return head.next

    # Second pass: find the node before the one to remove
    current = head
    for i in range(length - n - 1):
        current = current.next

    # Remove the nth node
    current.next = current.next.next
    return head


# Helper function to create a linked list from a list
def create_linked_list(values):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head


# Helper function to convert linked list to list for easy printing
def linked_list_to_list(head):
    result = []
    current = head
    while current:
        result.append(current.val)
        current = current.next
    return result


# Test cases
if __name__ == "__main__":
    # Test 1: [1, 2, 3, 4, 5], n=2
    head1 = create_linked_list([1, 2, 3, 4, 5])
    result1 = remove_nth_node_two_pass(head1, 2)
    print(linked_list_to_list(result1))  # [1, 2, 3, 5]

    # Test 2: [1], n=1
    head2 = create_linked_list([1])
    result2 = remove_nth_node_two_pass(head2, 1)
    print(linked_list_to_list(result2))  # []

    # Test 3: [1, 2], n=2
    head3 = create_linked_list([1, 2])
    result3 = remove_nth_node_two_pass(head3, 2)
    print(linked_list_to_list(result3))  # [2]

    # Test 4: [1, 2], n=1
    head4 = create_linked_list([1, 2])
    result4 = remove_nth_node_two_pass(head4, 1)
    print(linked_list_to_list(result4))  # [1]

File: solutions/Golang/remove_nth_node_from_end_of_list_two_pass.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

/*
Remove the nth node from the end of a linked list using two passes.

Approach:
1. First pass: count the total number of nodes
2. Second pass: find and skip the nth node from the end

Time: O(L) + O(L-n) = O(L)
Space: O(1)
*/
func removeNthNodeTwoPass(head *ListNode, n int) *ListNode {
    // First pass: count the total nodes
    length := 0
    current := head
    for current != nil {
        length++
        current = current.Next
    }

    // Edge case: removing the head node
    if length == n {
        return head.Next
    }

    // Second pass: find the node before the one to remove
    current = head
    for i := 0; i < length-n-1; i++ {
        current = current.Next
    }

    // Remove the nth node
    current.Next = current.Next.Next
    return head
}

// Helper function to create a linked list from a slice
func createLinkedList(values []int) *ListNode {
    if len(values) == 0 {
        return nil
    }
    head := &ListNode{Val: values[0]}
    current := head
    for i := 1; i < len(values); i++ {
        current.Next = &ListNode{Val: values[i]}
        current = current.Next
    }
    return head
}

// Helper function to convert linked list to slice for easy printing
func linkedListToSlice(head *ListNode) []int {
    var result []int
    current := head
    for current != nil {
        result = append(result, current.Val)
        current = current.Next
    }
    return result
}

// Test cases
func main() {
    // Test 1: [1, 2, 3, 4, 5], n=2
    head1 := createLinkedList([]int{1, 2, 3, 4, 5})
    result1 := removeNthNodeTwoPass(head1, 2)
    fmt.Println(linkedListToSlice(result1))  // [1 2 3 5]

    // Test 2: [1], n=1
    head2 := createLinkedList([]int{1})
    result2 := removeNthNodeTwoPass(head2, 1)
    fmt.Println(linkedListToSlice(result2))  // []

    // Test 3: [1, 2], n=2
    head3 := createLinkedList([]int{1, 2})
    result3 := removeNthNodeTwoPass(head3, 2)
    fmt.Println(linkedListToSlice(result3))  // [2]

    // Test 4: [1, 2], n=1
    head4 := createLinkedList([]int{1, 2})
    result4 := removeNthNodeTwoPass(head4, 1)
    fmt.Println(linkedListToSlice(result4))  // [1]
}
